src/System.Buffers.Experimental/System/Buffers/SpanWriter.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Buffers.Text;
+using System.Diagnostics;
+using System.Text;
+
+namespace System.Buffers
+{
+    public ref struct SpanWriter
+    {
+        Span<byte> _buffer;
+        int _written;
+        SymbolTable _symbols;
+
+        public ReadOnlyMemory<byte> Newline;
+        public Func<int, Memory<byte>> Enlarge;
+
+        static ReadOnlyMemory<byte> s_defaultNewline = new byte[] { (byte)'\n' };
+
+        public SpanWriter(Span<byte> buffer, SymbolTable symbols = default(SymbolTable))
+        {
+            _buffer = buffer;
+            _written = 0;
+            _symbols = symbols == default(SymbolTable) ? SymbolTable.InvariantUtf8 : symbols;
+            Newline = s_defaultNewline;
+            Enlarge = null;
+        }
+
+        public void WriteBytes<T>(T value, IBufferTransformation transformation = null) where T : IWritable
+        {
+            var start = Index;
+            int written;
+            while (!value.TryWrite(Free, out written, default(ParsedFormat)))
+            {
+                Resize();
+            }
+
+            if (transformation != null)
+            {
+                var status = transformation.Transform(Buffer.Slice(start), written, out written);
+                if (status == OperationStatus.Done)
+                {
+                    _written += written;
+                    return;
+                }
+
+                if (status == OperationStatus.InvalidData || status == OperationStatus.NeedMoreData) throw new Exception("invalid value or transformation");
+                // if OperationStatus.DestinationTooSmall
+                Resize();
+                WriteBytes(value, transformation);
```

Comment:
Instead of calling the function from the beginning again, why not go to line 41 (inside the `if (transformation != null)`) block)? We have already written the value into the buffer so there is no need to redo that.

Reply:
Because the transformation is in-place and could have destroyed the written value.